### PR TITLE
Allow to limit the length of a column in list view

### LINKED ISSUE #1073: make it possible to limit the maximum length of columns in the list view
something like `->add('body', 'text', array('length' => 100))`

### PR DESCRIPTION
**Pull request by nykopol**
This PR target the 3.x branch because it doesn't break compatibility

Fixes #1073 

## Changelog

add a length option to truncate columns on list view

```markdown
### Added
- add a length option to truncate columns on list view
```

## To do

- [x] Update the tests
- [x] Update the documentation

## Subject

This PR is a simple implementation to truncate columns on list view. I'm not 100% confident that it's the way to do it because I don't really understand where FieldDescription are defined.
Based on the propositions made in the issue and what I found in the code I have some questions.

* About using a script to hide/show the full value: this will not work if the value is in a link apart by splitting the link and the expand link. Do we split the links or we assume that this does not work on links ?
* About porting this functionality on show view, the template have a `safe` option that allow HTML value. The behaviour can not be implemented on this values. Otherwise, we will either have to remove HTML tags, or have the risk to break tags.

### CODE REVIEW

**Comment by nykopol** [comment]
@greg0ire Hi sorry but I can't find this discussion. How do you truncat with CSS ? With `text-overflow` the text is truncated when it can't fit in the element, but you have no control on the number of character to keep. The only solution that I see is with some JS.

**Comment by greg0ire** [comment]
Here is the discussion : https://github.com/sonata-project/SonataCoreBundle/pull/311

**Comment by greg0ire** [comment]
You're not the first to require this filter, so I'm not sure we should remove truncate anymore

**Comment by nykopol** [comment]
ok, I have read the discussion you pointed out. I understand better you comment. Maybe, I can get rid of this with a few lines of JS. It will avoid the difficult question about this filter and will be more user-friendly with a 'see more' link.

**Comment by greg0ire** [comment]
At the moment it might be the best thing to do because the discussion is kind of blocked. I just posted on the linked symfony discussion

**Comment by greg0ire** [comment]
For the broken test suite, see #4192 , please rebase once it is merged.

File: Resources/doc/reference/action_list.rst [hunk 1]
```diff
@@ -550,6 +556,23 @@ You can :
...
+
+.. code-block:: php
+
+            // ...
+            ->add('description', 'text', array(
+                'header_style' => 'width: 35%',
+                'collapse' => array(
+                    'height' => 40, // height in px
+                    'read_more' => 'I want to see the full description', // content of the "read more" link
+                    'read_less' => 'This text is to long, reduce the size' // content of the "read less" link
```
Comment (greg0ire) [inline Resources/doc/reference/action_list.rst]: To => too

**Comment by nykopol** [comment]
@greg0ire Thanks for you review. I've also added the option on the show view.

File: Resources/views/CRUD/base_list_field.html.twig [hunk 2]
```diff
@@ -25,10 +25,10 @@ file that was distributed with this source code.
...
                 {% if field_description.options.collapse is defined %}
                     {% set collapse = field_description.options.collapse %}
-                    <div class="sonata-truncate"
-                          data-truncate-height="{{collapse.height is defined ? collapse.height : 40}}"
-                          data-truncate-more="{{ (collapse.more is defined ? collapse.more : 'read_more')|trans({}, 'SonataAdminBundle') }}"
-                          data-truncate-less="{{ (collapse.less is defined ? collapse.less : 'read_less')|trans({}, 'SonataAdminBundle') }}">{{ value }}</div>
+                    <div class="sonata-readmore"
+                          data-readmore-height="{{ collapse.height is defined ? collapse.height : 40 }}"
+                          data-readmore-more="{{ (collapse.more is defined ? collapse.more : 'read_more')|trans({}, 'SonataAdminBundle') }}"
+                          data-readmore-less="{{ (collapse.less is defined ? collapse.less : 'read_less')|trans({}, 'SonataAdminBundle') }}">{{ value }}</div>
```
Comment (greg0ire) [inline Resources/views/CRUD/base_list_field.html.twig]: Can you please use the `default` filter here, while you're at it?

File: Resources/views/CRUD/base_show_field.html.twig [hunk 3]
```diff
@@ -10,7 +10,23 @@ file that was distributed with this source code.
...
-<td>{% block field %}{% if field_description.options.safe %}{{ value|raw }}{% else %}{{ value|nl2br }}{% endif %}{% endblock %}</td>
+<td>
+    {%- block field -%}
+        {% spaceless %}
+            {% if field_description.options.collapse is defined %}
+                {% set collapse = field_description.options.collapse %}
+                <div class="sonata-readmore"
+                      data-readmore-height="{{ collapse.height is defined ? collapse.height : 40 }}"
+                      data-readmore-more="{{ (collapse.more is defined ? collapse.more : 'read_more')|trans({}, 'SonataAdminBundle') }}"
+                      data-readmore-less="{{ (collapse.less is defined ? collapse.less : 'read_less')|trans({}, 'SonataAdminBundle') }}">
```
Comment (greg0ire) [inline Resources/views/CRUD/base_show_field.html.twig]: Same here, please use `default`

File: Resources/views/CRUD/base_list_field.html.twig [hunk 4]
```diff
@@ -9,7 +9,7 @@ file that was distributed with this source code.
 
 #}
 
-<td class="sonata-ba-list-field sonata-ba-list-field-{{ field_description.type }}" objectId="{{ admin.id(object) }}"{% if field_description.options.row_align is defined %} style="text-align:{{ field_description.options.row_align }}"{% endif %}>
+<td class="sonata-ba-list-field sonata-ba-list-field-{{ field_description.type }}" objectId="{{ admin.id(object) }}"{% if field_description.options.row_align is defined or field_description.options.cell_style is defined %} style="{{field_description.options.cell_style}}{% if field_description.options.row_align is defined %}text-align:{{ field_description.options.row_align }}{% endif %}"{% endif %}>
```
Comment (greg0ire) [inline Resources/views/CRUD/base_list_field.html.twig]: Please linebreak this line.

File: Resources/views/CRUD/base_show_field.html.twig [hunk 5]
```diff
@@ -10,7 +10,23 @@ file that was distributed with this source code.
...
+                {% set collapse = field_description.options.collapse %}
+                <div class="sonata-readmore"
+                      data-readmore-height="{{ collapse.height|default(40) }}"
+                      data-readmore-more="{{ collapse.more|default('read_more')|trans({}, 'SonataAdminBundle') }}"
+                      data-readmore-less="{{ collapse.less|default('read_less')|trans({}, 'SonataAdminBundle') }}">
+            {% endif %}
+
+            {% if field_description.options.safe %}{{ value|raw }}{% else %}{{ value|nl2br }}{% endif %}
+
+            {% if field_description.options.collapse is defined %}</div>{% endif %}
```
Comment (greg0ire) [inline Resources/views/CRUD/base_show_field.html.twig]: Repeating this condition feels weird. Maybe store the inner value in a variable and repeat it?

File: Resources/translations/SonataAdminBundle.de.xliff [hunk 6]
```diff
@@ -466,6 +466,14 @@
                 <source>label_unknown_user</source>
                 <target>unbekannt</target>
             </trans-unit>
+            <trans-unit id="read_more">
+                <source>read_more</source>
+                <target>read_more</target>
```
Comment (core23) [inline Resources/translations/SonataAdminBundle.de.xliff]: mehr anzeigen

File: Resources/translations/SonataAdminBundle.de.xliff [hunk 7]
```diff
@@ -466,6 +466,14 @@
...
+            </trans-unit>
+            <trans-unit id="read_less">
+                <source>read_less</source>
+                <target>read_less</target>
```
Comment (core23) [inline Resources/translations/SonataAdminBundle.de.xliff]: weniger anzeigen

**Comment by core23** [comment]
Why did you approve this @greg0ire ?

Tests are still missing

**Comment by nykopol** [comment]
@core23 thanks for your translation. I'll do the tests tomorrow.

**Comment by greg0ire** [comment]
> Why did you approve this @greg0ire ?

Oh right I overlooked this, sorry…

**Comment by nykopol** [comment]
@greg0ire @core23 The tests are updated. I think that now this PR is ready for merge.

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 8]
```diff
@@ -1215,6 +1215,31 @@ class="x-editable"
                     ),
                 ),
             ),
+
+            array(
+                '<td class="sonata-ba-list-field sonata-ba-list-field-text" objectId="12345">
+                <div class="sonata-readmore" data-readmore-height="40" data-readmore-more="Read more" data-readmore-less="Close">A very very very very very very very very very long string.</div>
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: Please try to keep all lines under 120 chars

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 9]
```diff
@@ -1218,20 +1218,22 @@ class="x-editable"
 
             array(
                 '<td class="sonata-ba-list-field sonata-ba-list-field-text" objectId="12345">
-                <div class="sonata-readmore" data-readmore-height="40" data-readmore-more="Read more" data-readmore-less="Close">A very very very very very very very very very long string.</div>
+                <div class="sonata-readmore" data-readmore-height="40" data-readmore-more="Read more" '.
+                'data-readmore-less="Close">A very very long string.</div>
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: Please use the heredoc notation instead, like this : 

```
<<<'EOT'
<div
    class="sonata-readmore"
    data-readmore-height="40"
    data-readmore-more="Read more"
    data-readmore-less="Close">
        A very very very very very very very very very long string
</div>
EOT
```

Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: There is [an example](https://github.com/nykopol/SonataAdminBundle/blob/526b2e731bfa817f8db83da3e9b2b7c9739990f8/Tests/Twig/Extension/SonataAdminExtensionTest.php#L656) in the file already

Comment (nykopol) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: Ok, but if i use the heredoc notation, it will break the limit of 120 characters.

Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: How so? The code snippet I just posted surely doesn't, right?

Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: With the indention it might, but I guess you can remove one "very"

**Comment by greg0ire** [comment]
Also, make sure to apply the StyleCI patch (see the "Details") link. After that, it should indeed be good to merge :)

File: Tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 10]
```diff
@@ -1217,23 +1217,27 @@ class="x-editable"
             ),
 
             array(
-                '<td class="sonata-ba-list-field sonata-ba-list-field-text" objectId="12345">
-                <div class="sonata-readmore" data-readmore-height="40" data-readmore-more="Read more" '.
-                'data-readmore-less="Close">A very very long string.</div>
-                </td>',
+                <<<'EOT'
+<td class="sonata-ba-list-field sonata-ba-list-field-text" objectId="12345">
+<div class="sonata-readmore" data-readmore-height="40" data-readmore-more="Read more" data-readmore-less="Close">a</div>
```
Comment (greg0ire) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: Why didn't you linebreak the attributes, just like I did in the code snippet I gave you?

Comment (nykopol) [inline Tests/Twig/Extension/SonataAdminExtensionTest.php]: I'm currently fighting against the `removeExtraWhitespace` method that i first didn't noticed. Actually, i have some spaces that are added or removed... I'm on it.

**Comment by greg0ire** [comment]
Reviewers : this should be squash-merged.

@nykopol : if you want to squash the commits you can, but don't feel obliged to.

**Comment by nykopol** [comment]
@greg0ire it's squashed

**Comment by greg0ire** [comment]
Great! @sonata-project/contributors , please give this a review and merge if it looks good.

File: Resources/doc/reference/action_list.rst [hunk 11]
```diff
@@ -525,6 +525,8 @@ You can :
 
 - `header_style`: Customize the style of header (width, color, background, align...)
 - `header_class`: Customize the class of the header
+- `cell_style`: Customize the style of inner cells (width, color, background, align...)
```
Comment (core23) [inline Resources/doc/reference/action_list.rst]: Would prefer a `cell_class` instead to keep html clean and do the styling with css.

Comment (nykopol) [inline Resources/doc/reference/action_list.rst]: I added this for convenience following the existing option `header_style`. But if it has to be discussed, i think that it's better to do this in a other PR. This is not the purpose of this one, so i removed this.

Comment (core23) [inline Resources/doc/reference/action_list.rst]: IMHO, the `header_style` option wasn't a good idea at all. 

wdyt @greg0ire @OskarStark  ? Should we deprecate / drop it?

Comment (greg0ire) [inline Resources/doc/reference/action_list.rst]: I think so. It gives flexibility, but that's quite ugly. Not sure who uses that…

**Comment by greg0ire** [comment]
Thanks @nykopol !

**Pull request merged**